### ISSUE TITLE
Since 2023/12/04: Unable to refresh the access token

### ISSUE THREAD

**Issue by daiginjo** [guidone/node-red-contrib-netatmo-dashboard#28]
Since today I've alway get the following error:
`[netatmo-dashboard:b9345e53.83ab5] Unable to refresh the access token`

When I generate a new refresh token, enter it in node-red, one request works. From the second call, again the error above.
Something changed by netatmo?

**Comment by mrmaximas** [comment]
Same sh*t.

**Comment by daiginjo** [comment]
Found out, that the error is a: `{"error":"invalid_grant"}`
but why?

**Comment by mrmaximas** [comment]
anyone using api in two places at once like me? homebridge + node-red?

**Comment by DConcord** [comment]
Although I haven't found any official statements, it looks like Netatmo is now refreshing BOTH the access token AND the refresh token when token is refreshed (so the previous refresh token is now invalid). Standard Oath2 behavior, but a breaking change for Netatmo integrations

**Comment by daiginjo** [comment]
Yes, seems to be the cause of this issue. Already mentioned here: https://forum.netatmo.com/viewtopic.php?p=95627#p95627

**Comment by mrmaximas** [comment]
@guidone 

Copypasted from https://github.com/skrollme/homebridge-eveatmo/issues/77#issuecomment-1840885487

Info from Netatmo DEV Support:

We just did a modification on the token retrieval process :
 
When you refreshed an access_token using the associated endpoint [https://api.netatmo.com/oauth2/token, Netatmo servers](https://api.netatmo.com/oauth2/token,%C2%A0Netatmo%C2%A0servers) responded with a couple of tokens : an access_token and a refresh_token. If the previous access_token was still valid, the refresh_token value was never renewed
 
Starting from the 04/12/2023, this behavior changed to be compliant with the recommendations of the [RFC of the OAuth2 Authorization Framework](https://www.rfc-editor.org/rfc/rfc6749) (section 10.4) and improving the security of the data of our users
When refreshing tokens, access_token and refresh_token values will be automatically renewed and former tokens invalidated
 
So, if you do not update and use the new refresh_token value when refreshing your access_token, your users will be disconnected after 3 hours and you will retrieve an “invalid_grant” error
To fix it, you need to update the tokens value as soon as you get the newly generated ones.

**Comment by fmarzocca** [comment]
> To fix it, you need to update the tokens value as soon as you get the newly generated ones.

Does this mean that you have to refresh the tokens every 3 hours?

**Comment by jostrasser** [comment]
Related to https://github.com/guidone/node-red-contrib-netatmo-dashboard/issues/25

**Comment by jostrasser** [comment]
> > To fix it, you need to update the tokens value as soon as you get the newly generated ones.
> 
> Does this mean that you have to refresh the tokens every 3 hours?

I´ll try to get a confirmation from Netatmo.

**Comment by mrmaximas** [comment]
> > To fix it, you need to update the tokens value as soon as you get the newly generated ones.
> 
> Does this mean that you have to refresh the tokens every 3 hours?

every reading cycle

**Comment by jostrasser** [comment]
Additional info from Netatmo:

```
The access_token must be refreshed every 3 hours (10800 seconds). This has always been the case
The difference was that if you refreshed your access_token before its expiration, refresh_token value always remained the same. If you wanted to refresh it but after expiration, the refresh_token value changed
 
Now, the 2 values change at each new /token request. So, you must catch the new generated refresh_token value each time a new request is done to the endpoint, and perform the next /token call with this new value
```

**Comment by ThomasWilla** [comment]
Here is an alternative Flow with OAuth2 and refreshing the refresh token every 60 minutes. It works 
https://gist.github.com/matz33/843165d0046cac534864b65e633801a5#file-flow-json

**Comment by jostrasser** [comment]
> Here is an alternative Flow with OAuth2 and refreshing the refresh token every 60 minutes. It works https://gist.github.com/matz33/843165d0046cac534864b65e633801a5#file-flow-json

Cool, thank you!
Do you have any idea, why it is not possible to get the detailed data with your workflow?

If I use the node-red plugin, I am getting more data from the Rain Gauge, also the 1hour and 24hrs data:

```
[msg.payload.detailed.body.devices[0].home_name+"_Regensensor_Rain"]: msg.payload.detailed.body.devices[0].modules[2].dashboard_data.Rain,
    [msg.payload.detailed.body.devices[0].home_name+"_Regensensor_Sum_Rain_1"]: msg.payload.detailed.body.devices[0].modules[2].dashboard_data.sum_rain_1,
    [msg.payload.detailed.body.devices[0].home_name+"_Regensensor_Sum_Rain_24"]: msg.payload.detailed.body.devices[0].modules[2].dashboard_data.sum_rain_24,
```

Thanks!

**Comment by jostrasser** [comment]
> Here is an alternative Flow with OAuth2 and refreshing the refresh token every 60 minutes. It works https://gist.github.com/matz33/843165d0046cac534864b65e633801a5#file-flow-json

@ThomasWilla Cool, thanks so much! :)

**Comment by ThomasWilla** [comment]
Hello Johannes,

I get the value for rain sum 1 and 24 also with the “new” implementation.


Regards, Thomas

**Comment by jostrasser** [comment]
Hi @ThomasWilla ... It was my issue, sorry!
I can confirm that the mentioned workflow is running really stable for me! :) 
Thanks!

**Comment by jostrasser** [comment]
Another update from Netatmo. 
This is the reason why all the apps, plugins, integrations are back working again:

`Indeed, teams had to do a rollback to fix a problem for multiple connections (if a 2nd device connects, it asks for a token and invalidates the access for the 1st connected device). I reported that it was a problem for third-party apps
The plan is still to put in place this change. They are working on a solution, but I don't have for now more information about a release date`

I requested detailed information when the next change will take place.

**Comment by S474N** [comment]
From where you have NetatmoAccessToken? I have only NetatmoClientId and NetatmoClientSecret.

**Comment by jostrasser** [comment]
> From where you have NetatmoAccessToken? I have only NetatmoClientId and NetatmoClientSecret. 

I am creating the Access and Refresh token with Postman. I described it here:

https://github.com/skrollme/homebridge-eveatmo/issues/62#issuecomment-1260603695

Or you can use any other OAuth2 client to create it. Check out the Readme / Authentication: https://github.com/guidone/node-red-contrib-netatmo-dashboard#authentication

**Comment by kevinpato72** [comment]
I received a mail from Netatmo, they will reactivate the feature in May.

**Comment by FoeMaz** [comment]
So they did it, connection stopped working on June 3rd between 11 AM and 12 PM CEST.
Any insights if this is fixable?

**Comment by S474N** [comment]
can you try to create a fresh token pair?
See: https://gist.github.com/matz33/843165d0046cac534864b65e633801a5?permalink_comment_id=5078029#gistcomment-5078029

**Comment by FoeMaz** [comment]
I'm trying to refresh using the oath2 flow, but I always get "invalid_request"
![grafik](https://github.com/guidone/node-red-contrib-netatmo-dashboard/assets/84348209/280f692d-d277-4928-9ed8-bb754a31a952)

is there something obvious I'm missing?

**Comment by martiko70** [comment]
For several month, everything worked fine, now I keep getting the error "Unable to refresh the access token". I already created new access and refresh token with RapidAPI (which worked fine) but even when I paste the new refresh token in the node-configuration, the error persists...

Any idea what I am doing wrong? Or did Netatmo change something again?

**Comment by jostrasser** [comment]
> can you try to create a fresh token pair?
> 
> See: https://gist.github.com/matz33/843165d0046cac534864b65e633801a5?permalink_comment_id=5078029#gistcomment-5078029

This is not enough for this plugin. The plugin cannot handle the OAuth2 change. 

See: https://github.com/guidone/node-red-contrib-netatmo-dashboard/issues/25#issuecomment-2147194044

**Comment by daiginjo** [comment]
After resetting the token pair, just one successful measurement fetch is working fine. Then same issue as described above.

**Comment by FoeMaz** [comment]
> can you try to create a fresh token pair? See: https://gist.github.com/matz33/843165d0046cac534864b65e633801a5?permalink_comment_id=5078029#gistcomment-5078029

Seems to have worked, I got the token pair from the netatmo site and then I got it up and running again

**Comment by mtressl** [comment]
It seems it only works for one request, and then the refresh token becomes invalid again.

**Comment by cheinelt** [comment]
Same here: I get a new refresh token from the netatmo app page. After I have entered that fresh refresh token in the netatmo-config-node, it gets the netatmo station's data once, but on every further request, it returns "Unable to refresh the access token". Any plans to fix this? Would be very much appreciated!

**Comment by kailausberg** [comment]
> Same here: I get a new refresh token from the netatmo app page. After I have entered that fresh refresh token in the netatmo-config-node, it gets the netatmo station's data once, but on every further request, it returns "Unable to refresh the access token". Any plans to fix this? Would be very much appreciated!

If you read all the comments here, you will find out, that this git is not maintained and not working anymore...  A great alternative is mentioned above.

**Comment by mdisabato** [comment]
I followed the directions and it's working fine. Survived a restart of Home Assistant and the NodeRED container.
